File: java/src/main/java/org/revenj/server/commands/SubmitEvent.java
package org.revenj.server.commands;

import org.revenj.patterns.*;
import org.revenj.server.CommandResult;
import org.revenj.server.ServerCommand;

import java.io.IOException;
import java.util.Optional;

public class SubmitEvent implements ServerCommand {

	private final DomainModel domainModel;

	public SubmitEvent(DomainModel domainModel) {
		this.domainModel = domainModel;
	}

	public static final class Argument<TFormat> {
		public String Name;
		public TFormat Data;
		public Boolean ReturnInstance;

		public Argument(String name, TFormat data, Boolean returnInstance) {
			this.Name = name;
			this.Data = data;
			this.ReturnInstance = returnInstance;
		}

		@SuppressWarnings("unused")
		private Argument() {
		}
	}

	@Override
	public <TInput, TOutput> CommandResult<TOutput> execute(ServiceLocator locator, Serialization<TInput> input, Serialization<TOutput> output, TInput data) {
		Argument<TInput> arg;
		try {
			arg = input.deserialize(data, Argument.class, data.getClass());
		} catch (IOException e) {
			return CommandResult.badRequest(e.getMessage());
		}
		Optional<Class<?>> manifest = domainModel.find(arg.Name);
		if (!manifest.isPresent()) {
			return CommandResult.badRequest("Unable to find specified domain object: " + arg.Name);
		}
		if (arg.Data == null) {
			return CommandResult.badRequest("Data to submit not specified.");
		}
		DomainEvent instance;
		try {
			instance = (DomainEvent) input.deserialize(manifest.get(), arg.Data);
		} catch (IOException e) {
			return CommandResult.badRequest("Error deserializing provided input for: " + arg.Name + ". Reason: " + e.getMessage());
		}
		DomainEventStore store;
		try {
			store = locator.resolve(DomainEventStore.class, manifest.get());
		} catch (ReflectiveOperationException e) {
			return CommandResult.badRequest("Error resolving event store for: " + arg.Name + ". Reason: " + e.getMessage());
		}
		String uri = store.submit(instance);
		return new CommandResult<>(Boolean.TRUE.equals(arg.ReturnInstance) ? output.serialize(store) : output.serialize(uri), "Event stored", 201);
	}
}

File: scala/revenj-akka/src/main/scala/net/revenj/server/commands/SubmitEvent.scala
package net.revenj.server.commands

import net.revenj.patterns._
import net.revenj.serialization.Serialization
import net.revenj.server.commands.SubmitEvent.Argument
import net.revenj.server.{CommandResult, ServerCommand}

import scala.concurrent.Future

class SubmitEvent(domainModel: DomainModel) extends ServerCommand {

  override def execute[TInput, TOutput](
    locator: ServiceLocator,
    input: Serialization[TInput],
    output: Serialization[TOutput],
    data: TInput): Future[CommandResult[TOutput]] = {

    val arg = input.deserializeRuntime(data, classOf[Argument[TInput]], data.getClass)
    lazy val manifest = domainModel.find(arg.get.Name)
    if (!arg.isSuccess) {
      CommandResult.badRequest(arg.failed)
    } else if (manifest.isEmpty) {
      CommandResult.badRequest(s"Unable to find specified domain object: ${arg.get.Name}")
    } else if (arg.get.Data == null) {
      CommandResult.badRequest("Data to submit not specified")
    } else if (!classOf[DomainEvent].isAssignableFrom(manifest.get)) {
      CommandResult.badRequest(s"Specified type is not an domain event: ${arg.get.Name}")
    } else {
      val instance = input.deserializeRuntime[DomainEvent](arg.get.Data, manifest.get)
      if (!instance.isSuccess) {
        CommandResult.badRequest(s"Error deserializing provided input for: ${arg.get.Name}. Reason: ${instance.failed.get.getMessage}")
      } else {
        val tryStore = locator.resolve(classOf[DomainEventStore[DomainEvent]], manifest.get)
        if (!tryStore.isSuccess) {
          CommandResult.badRequest(s"Error resolving event store for: ${arg.get.Name}. Reason: ${tryStore.failed.get.getMessage}")
        } else {
          import scala.concurrent.ExecutionContext.Implicits.global
          tryStore.get.submit(instance.get).map { uri =>
            val returnInstance = arg.get.ReturnInstance.getOrElse(false)
            val response = output.serializeRuntime(if (returnInstance) instance.get else uri)
            if (response.isSuccess) {
              CommandResult[TOutput](Some(response.get), "Event stored", 201)
            } else {
              CommandResult[TOutput](None, response.failed.get.getMessage, 500)
            }
          }
        }
      }
    }
  }
}

object SubmitEvent {
  case class Argument[TFormat](
    Name: String,
    Data: TFormat,
    ReturnInstance: Option[Boolean]
  )
}